### PR TITLE
Adding ADE7953 irq_pin

### LINKED ISSUE #1529: Shelly 2.5 with ADE7953 getting hot (76ºC)
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Docker

**ESP (ESP32/ESP8266, Board/Sonoff):**

Shelly 2.5

**ESPHome version (latest production, beta, dev branch)**

v1.15.2

**Affected component:**

ade7953

**Description of problem:**
The device is getting hot with all configurations (including an empty one). The internal NTC is measuring 76ºC. It is also hot to the touch.

**Problem-relevant YAML-configuration entries:**
```yaml
# Basic Config
substitutions:
  devicename: shelly_25


esphome:
  name: ${devicename}
  platform: ESP8266
  board: esp01_1m
 
i2c:
  sda: GPIO12
  scl: GPIO14

    
  # NTC Temperature
  - platform: ntc
    sensor: temp_resistance_reading
    name: ${devicename} Temperature
    unit_of_measurement: "°C"
    accuracy_decimals: 1
    icon: "mdi:thermometer"
    calibration:
      b_constant: 3350
      reference_resistance: 10kOhm
      reference_temperature: 298.15K
  - platform: resistance
    id: temp_resistance_reading
    sensor: temp_analog_reading
    configuration: DOWNSTREAM
    resistor: 32kOhm
    internal: true
  - platform: adc
    id: temp_analog_reading
    pin: A0
    update_interval: 30s

# ADDING THIS FIXES THE PROBLEM
binary_sensor:
  - platform: gpio
    pin:
      number: GPIO16
    name: ${devicename} IRQ?

```

**Logs (if applicable):**

```


```

**Additional information and things you've tried:**

The problem seems to be that in this device the ADE7953 IRQ pin is connected to the GPIO16. The ade7953 component does not initialize this pin as an input, so it actually acts some kind of load overheating the esp chip. 

I don't' know any other devices with a similar setup. Not sure if make sense to modify the ade7953 to handle this or not.

I've tried adding a dummy binary sensor to the GPIO16 and the problem disappeared, lowering the temperature 20 degrees. However, it's a very weird solution.

### PR DESCRIPTION
**Pull request by djwmarcx**
## Description:

**Related issue :** fixes https://github.com/esphome/issues/issues/1529

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation:** esphome/esphome-docs#848

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ade7953/ade7953.h [hunk 1]
```diff
@@ -9,6 +9,10 @@ namespace ade7953 {
 
 class ADE7953 : public i2c::I2CDevice, public PollingComponent {
  public:
+  void set_irq_pin(uint8_t irq_pin) {
+    has_irq_ = true;
+    irq_pin_number_ = irq_pin;
+  }
```
Comment (jesserockz) [inline esphome/components/ade7953/ade7953.h:15]:
```suggestion
  void set_irq_pin(GPIOPin *irq_pin) {
    this->irq_pin_ = irq_pin;
  }
```

File: esphome/components/ade7953/ade7953.cpp [hunk 2]
```diff
@@ -8,6 +8,9 @@ static const char *TAG = "ade7953";
 
 void ADE7953::dump_config() {
   ESP_LOGCONFIG(TAG, "ADE7953:");
+  if (this->has_irq_) {
+    ESP_LOGCONFIG(TAG, "  IRQ Pin: GPIO%u", this->irq_pin_number_);
+  }
```
Comment (jesserockz) [inline esphome/components/ade7953/ade7953.cpp:13]:
```suggestion
  if (this->irq_pin_ != nullptr) {
    LOG_PIN("  IRQ Pin: ", this->irq_pin_);
  }
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
This code can be a lot simpler. Take a look below at my suggestions =)

**Comment by djwmarcx** [comment]
Thanks for your suggestions, Jesse!

I understand the reason are you proposing those changes. It was my initial approach actually. The problem is that the IRQ pin in the ADE7953 IRQ pin is pulled to HIGH and using the full pin schema can lead to someone setting the pin as INPUT_PULLDOWN causing the overheating problem I'm trying to avoid with my changes. In the Shelly 2.5 specifically, this is not possible because the GPIO doesn't have internal pull down resistor, but in different setups

**Review by jesserockz — APPROVED** [review]
Ok we can leave it as is.

**Pull request merged**